File: streamviewer.py
#!/usr/bin/env python3

import sys

from PyQt5.QtCore import Qt, QUrl, pyqtSignal
from PyQt5.QtWidgets import QApplication, QWidget, QGridLayout, QStackedLayout, QLabel
from PyQt5.QtGui import QMovie, QPalette
from PyQt5.QtMultimedia import QMediaPlayer, QMediaContent
from PyQt5.QtMultimediaWidgets import QVideoWidget


class StreamViewer(QWidget):
    streams_ready = pyqtSignal()

    def __init__(self, streams):
        self.streams = streams
        self.starting_player_indices = set(range(len(self.streams)))
        super().__init__()

        self.setAttribute(Qt.WA_QuitOnClose)
        self.setWindowFlags(Qt.FramelessWindowHint)

        palette = QPalette()
        palette.setColor(QPalette.Background, Qt.white)
        self.setPalette(palette)

        self.layout = QStackedLayout()
        self.setLayout(self.layout)

        label = QLabel()
        label.setAlignment(Qt.AlignCenter)
        self.loading_gif = QMovie("loading.gif")
        label.setMovie(self.loading_gif)
        self.layout.insertWidget(0, label)

        self.streams_ready.connect(self.showStartingScreen)
        self.players = []
        for stream in self.streams:
            player = QMediaPlayer()
            player.setMedia(QMediaContent(stream))
            player.mediaStatusChanged.connect(self.changedMediaStatus)
            player.play()
            self.players.append(player)

        grid_layout = QGridLayout()
        grid_layout.setSpacing(0)
        grid_layout.setContentsMargins(0, 0, 0, 0)
        self.grid_widget = QWidget()
        self.grid_widget.setLayout(grid_layout)

        i = 0
        self.grid_widgets = []
        for row in range(2):
            for column in range(2):
                video_widget = QVideoWidget()
                self.players[i].setVideoOutput(video_widget)
                self.grid_widgets.append(video_widget)
                grid_layout.addWidget(video_widget, row, column)
                i+=1
        self.layout.insertWidget(1, self.grid_widget)

        video_widget = QVideoWidget()
        self.players[4].setVideoOutput(video_widget)
        self.layout.insertWidget(2, video_widget)

        self.loading_gif.start()

    def showStartingScreen(self):
        self.showStreamGrid()
        self.loading_gif.stop()
        for player in self.players:
            player.mediaStatusChanged.disconnect(self.changedMediaStatus)

    def showStreamFullScreen(self, index):
        if index > 3:
            self.layout.setCurrentIndex(index - 2)
        else:
            self.grid_widgets[index].show()
            for i, widget in enumerate(self.grid_widgets):
                if i != index:
                    widget.hide()
            self.layout.setCurrentIndex(1)

    def showStreamGrid(self):
        for widget in self.grid_widgets:
            widget.show()
        self.layout.setCurrentIndex(1)

    def keyPressEvent(self, event):
        key = event.key()
        if key == Qt.Key_Escape:
            self.close()
        elif key in (Qt.Key_1, Qt.Key_2, Qt.Key_3, Qt.Key_4, Qt.Key_5):
            self.showStreamFullScreen(key - Qt.Key_1)
        elif key == (Qt.Key_0):
            self.showStreamGrid()

    def isStreamReady(self, player):
        return (player.mediaStatus() in (QMediaPlayer.BufferedMedia,
                                         QMediaPlayer.StalledMedia) or
                player.state() == QMediaPlayer.StoppedState)

    def changedMediaStatus(self, status):
        player = self.sender()
        index = self.players.index(player)
        if index in self.starting_player_indices and self.isStreamReady(player):
            self.starting_player_indices.remove(index)

        if not self.starting_player_indices:
            self.streams_ready.emit()


STREAM_IDS = [
    '04asu087t_1@409528',
    '074tzyqju_1@394719',
    '0i1p9ue46_1@409669',
    '04et6qkjl_1@409668',
    '0xorfu8d8_1@308855'
]

STREAM_URL_FMT = "http://streamamghdl-lh.akamaihd.net/i/{}/index_1_av-p.m3u8"

streams = [QUrl(STREAM_URL_FMT.format(id)) for id in STREAM_IDS]

app = QApplication(sys.argv)

viewer = StreamViewer(streams)
viewer.showFullScreen()

app.exec_()

File: streamviewer.cpp
#include "streamviewer.h"

StreamViewer::StreamViewer(const std::vector<QUrl>& streams)
  : layout(new QStackedLayout), streams(streams), numStreams(streams.size())
{
  setAttribute(Qt::WA_QuitOnClose);
  setWindowFlags(Qt::FramelessWindowHint);
  setLayout(layout);

  QPalette *palette = new QPalette;
  palette->setColor(QPalette::Background, Qt::white);
  setPalette(*palette);

  QLabel *label = new QLabel;
  label->setAlignment(Qt::AlignCenter);
  loading = new QMovie(":loading.gif");
  label->setMovie(loading);
  layout->insertWidget(0, label);

  for (const QUrl &stream : streams) {
    QMediaPlayer *player = new QMediaPlayer;
    player->setMedia(stream);
    player->play();
    players.push_back(player);
    connect(player, &QMediaPlayer::mediaStatusChanged, this, &StreamViewer::checkIfStreamsReady);
  }

  QGridLayout *gridLayout = new QGridLayout;
  gridLayout->setSpacing(0);
  gridLayout->setContentsMargins(0, 0, 0, 0);
  QWidget *gridWidget = new QWidget();
  gridWidget->setLayout(gridLayout);
  layout->insertWidget(1, gridWidget);

  int i = 0;
  for (int row=0; row<2; row++) {
    for (int column=0; column<2; column++) {
      QVideoWidget *gridVideoWidget = new QVideoWidget;
      gridVideoWidgets.push_back(gridVideoWidget);
      players[i]->setVideoOutput(gridVideoWidget);
      gridLayout->addWidget(gridVideoWidget, row, column);
      i++;
    }
  }

  for (int i=4; i<numStreams; i++) {
    QVideoWidget *videoWidget = new QVideoWidget;
    players[i]->setVideoOutput(videoWidget);
    layout->insertWidget(i - 2, videoWidget);
  }

  connect(this, SIGNAL(streamsReady()), this, SLOT(showStartingScreen()));
  loading->start();
}

void StreamViewer::showStartingScreen()
{
  showStreamGrid();
  loading->stop();
}

void StreamViewer::checkIfStreamsReady(QMediaPlayer::MediaStatus state)
{
  Q_UNUSED(state);
  for (const auto &player : players) {
    if (player->mediaStatus() == QMediaPlayer::BufferedMedia ||
        player->mediaStatus() == QMediaPlayer::StalledMedia ||
        player->state() == QMediaPlayer::StoppedState) {
      disconnect(player, &QMediaPlayer::mediaStatusChanged, this, &StreamViewer::checkIfStreamsReady);
    } else return;
  }
  emit streamsReady();
}

void StreamViewer::showStreamGrid()
{
  for (int i=0; i<4; i++) {
    gridVideoWidgets[i]->show();
  }
  layout->setCurrentIndex(1);
}

void StreamViewer::showStreamFullScreen(int index)
{
  if (index > 3) layout->setCurrentIndex(index - 2);
  else {
    gridVideoWidgets[index]->show();
    for (int i=0; i<4; i++) {
      if (i != index) {
        gridVideoWidgets[i]->hide();
      }
    }
    layout->setCurrentIndex(1);
  }
}

void StreamViewer::keyPressEvent(QKeyEvent *event)
{
  int key = event->key();
  int keyIndex = key - Qt::Key_1;
  if (keyIndex >= 0 && keyIndex < numStreams) showStreamFullScreen(keyIndex);
  else {
    switch(key) {
      case Qt::Key_0:
      case Qt::Key_F11:
        showStreamGrid();
        break;
      case Qt::Key_Escape:
        for (QMediaPlayer* player : players) player->stop();
        close();
        break;
    }
  }
}
